### PR TITLE
stdlib: add Array.fold_left_map

### PR DESCRIPTION
**Pull request by craigfe**
Found myself wanting an `Array` equivalent of `List.fold_left_map` (added in 4.11.0), so decided to propose adding it. The documentation comments and tests are very similar to the existing `List` equivalents.

### CODE REVIEW

**Comment by craigfe** [comment]
Rebased. I'd appreciate a review on this.

File: stdlib/arrayLabels.mli
Comment (damiendoligez) [inline stdlib/arrayLabels.mli]: This line is too long and check-typo complains.

**Review by damiendoligez — APPROVED** [review]
Looks good and the code is obviously correct, but now I'm left wanting:
* more explicit documentation (who added `List.fold_left_map` anyway? @bschommer maybe, with the complicity of @alainfrisch?)
* the same function in `Seq`, `Bytes`, `Float.Array`, and maybe even `Set`, `Map`, `Queue`, `Stack`, `Weak`...

Good to go, after you split the over-long line.

**Comment by craigfe** [comment]
Thanks @damiendoligez for the review. I've fixed the overly-long line in question.

Re. providing this function elsewhere, I agree that it's worthwhile. I'm happy to do so myself (along with making the documentation more explicit) if there is support among maintainers for this.

On a related note, I've been slowly creating a spreadsheet that catalogues some of these inconsistencies (for instance, we have `{List,Array}.init` but not `Seq.init`). If there's support for solving those issues in a single pass – rather than adding them piece-wise as is done currently – I'm happy to invest the necessary time into PRs.

**Comment by bschommer** [comment]
> Looks good and the code is obviously correct, but now I'm left wanting:
> 
> * more explicit documentation (who added `List.fold_left_map` anyway? @bschommer maybe, with the complicity of @alainfrisch?)

The documentation is kind of the same as Batteries and Base use.

> * the same function in `Seq`, `Bytes`, `Float.Array`, and maybe even `Set`, `Map`, `Queue`, `Stack`, `Weak`...

For other data structures it is not that natural to add, @alainfrisch mentioned in https://github.com/ocaml/ocaml/pull/8894#issuecomment-546251911:

> I find it much less natural for data structures that implements a map but which don't have a canonical traversal ordering. E.g. Map is borderline (its map function specifies the traversal ordering, so that would be ok-ish), but if we had an Hashtbl.map (copying an hashtable, mapping values, with an unspecified traversal ordering), that wouldn't be so good.

**Comment by damiendoligez** [comment]
>On a related note, I've been slowly creating a spreadsheet that catalogues some of these inconsistencies (for instance, we have {List,Array}.init but not Seq.init). If there's support for solving those issues in a single pass – rather than adding them piece-wise as is done currently – I'm happy to invest the necessary time into PRs.

If you lump them together in one PR, the discussion will probably devolve into endless bike-shedding. OTOH it would be good to have a high-level view of the situation before discussing each addition in detail, so you could create a meta-issue to support the discussion.

**Pull request merged**
